iroh-net/src/discovery/local_node_discovery.rs

```diff
@@ -0,0 +1,255 @@
+                        ) {
+                            Ok(guard) => Some(guard),
+                            Err(e) => {
+                                tracing::error!(
+                                    "LocalNodeDiscovery error creating discovery service: {e}"
+                                );
+                                return;
+                            }
+                        };
+                    }
+                }
+            }
+        });
+        Self {
+            handle,
+            sender: send.clone(),
+        }
+    }
+
+    fn spawn_discoverer(
+        node_id: PublicKey,
+        sender: Sender<Message>,
+        socketaddrs: BTreeSet<SocketAddr>,
+        rt: &tokio::runtime::Handle,
+    ) -> Result<DropGuard> {
+        let callback = move |node_id: &str, peer: &Peer| {
+            tracing::trace!(
+                node_id,
+                ?peer,
+                "Received peer information from LocalNodeDiscovery"
+            );
+            sender
+                .send(Message::Discovery((node_id.to_string(), peer.clone())))
+                .ok();
+        };
+
+        let mut addrs: HashMap<u16, Vec<IpAddr>> = HashMap::default();
+        let socketaddrs: BTreeSet<SocketAddr> = socketaddrs
+            .into_iter()
+            .filter(|socketaddr| socketaddr.is_ipv4())
+            .collect();
+        for socketaddr in socketaddrs {
+            addrs
+                .entry(socketaddr.port())
+                .and_modify(|a| a.push(socketaddr.ip()))
+                .or_insert(vec![socketaddr.ip()]);
+        }
+
+        let mut discoverer =
+            Discoverer::new_interactive(N0_MDNS_SWARM.to_string(), node_id.to_string())
+                .with_callback(callback)
+                .with_ip_class(IpClass::V4Only);
+        if !addrs.is_empty() {
+            for addr in addrs {
+                discoverer = discoverer.with_addrs(addr.0, addr.1);
+            }
+        }
+        discoverer.spawn(rt)
+    }
+}
+
+impl From<&Peer> for DiscoveryItem {
+    fn from(peer_info: &Peer) -> Self {
+        let direct_addresses: BTreeSet<SocketAddr> = peer_info
+            .addrs()
+            .iter()
+            .map(|(ip, port)| SocketAddr::new(*ip, *port))
+            .collect();
+        DiscoveryItem {
+            provenance: PROVENANCE,
+            last_updated: None,
+            addr_info: AddrInfo {
+                relay_url: None,
+                direct_addresses,
+            },
+        }
+    }
+}
+
+impl Discovery for LocalNodeDiscovery {
+    fn resolve(&self, _ep: Endpoint, node_id: NodeId) -> Option<BoxStream<Result<DiscoveryItem>>> {
+        let (send, recv) = flume::bounded(20);
+        self.sender.send(Message::SendAddrs((node_id, send))).ok();
+        Some(recv.into_stream().boxed())
+    }
+
+    fn publish(&self, info: &AddrInfo) {
+        self.sender
+            .send(Message::ChangeLocalAddrs(info.clone()))
+            .ok();
+    }
+}
```

Comment:
can we have a simple unit like test that at least tests this on the same machine?